art/defences/preprocessor/preprocessor.py

```diff
@@ -21,10 +21,14 @@
 from __future__ import absolute_import, division, print_function, unicode_literals
 
 import abc
-from typing import List, Optional, Tuple
+from typing import List, Optional, Tuple, TYPE_CHECKING
 
 import numpy as np
 
+if TYPE_CHECKING:
+    import torch
+    import tensorflow as tf
+
```

Comment:
I am not sure if when you import torch and tensorflow at the same time, it may make some confict on gpus if they are available becaust tf will occupy all gpus by default, and then torch doesn't have one to use. However, here is just a type check and maybe it won't effect but I think you should check.

Reply:
In general for productions, I agree with your concerns, but as you mention these imports will only be carried out during the type-checking with mypy and so far we have not observed any problems.